From 1e70d63a9439f48b5f1a70accac531a10f4e4239 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Daniel=20Vr=C3=A1til?= <dvratil@kde.org>
Date: Mon, 10 Oct 2016 00:44:58 +0200
Subject: [PATCH] Create AgePostingSource on heap

There was an undocumented behaviour change in Xapian 1.4 where
Xapian::Query() no longer internally creates a clone of the
PostingResource that we pass to it and instead takes a (shared)
ownership of the pointer that is then re-used later while
the actual query is being executed, which means that the
PostingResource must live at least until the query execution
is finished.

We were creating the AgePostingSource on stack, which lead to
use-after-free in Xapian 1.4.

BUG: 363741
FIXED-IN: 5.3.2
---
 search/email/emailsearchstore.cpp | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/search/email/emailsearchstore.cpp b/search/email/emailsearchstore.cpp
index f108778..49be40a 100644
--- a/search/email/emailsearchstore.cpp
+++ b/search/email/emailsearchstore.cpp
@@ -117,7 +117,6 @@ QString EmailSearchStore::text(int queryId)
 
 Xapian::Query EmailSearchStore::finalizeQuery(const Xapian::Query &query)
 {
-    AgePostingSource ps(0);
-    return Xapian::Query(Xapian::Query::OP_AND_MAYBE, query, Xapian::Query(&ps));
+    return Xapian::Query(Xapian::Query::OP_AND_MAYBE, query, Xapian::Query(new AgePostingSource(0)));
 }
 
-- 
2.9.3

